icu4c/source/common/uresbund.cpp

```diff
@@ -1146,6 +1146,17 @@ static UResourceBundle *init_resb_result(const ResourceData *rdata, Resource r,
                                     r = res_findResource(&(dataEntry->fData), r, &myPath, &temp);
                                     if(r != RES_BOGUS) { /* found a resource, but it might be an indirection */
                                         resB = init_resb_result(&(dataEntry->fData), r, temp, -1, dataEntry, result, noAlias+1, resB, status);
+                                        if (temp == NULL || uprv_strcmp(keyPath, temp) != 0) {
+                                            // the call to init_resb_result() above will set resB->fKeyPath to be
```

Comment:
question: is that behavior a bug? or do we need it for something?

Reply:
You're asking whether setting both `fKey` and `fResPath` to the same thing is a bug?  I'm not completely sure, to tell you the truth, but I think it's generally desirable behavior other functions are generally relying on.  I think a lot of the time, in a lot of contexts, they actually _are_ the same, or it doesn't really matter.  But I did run into two spots (see ICU-20400 for the other one) where we really do want them to be different.  I thought about expanding `init_resb_result()` so that it just took a separate path parameter and set both things independently, but since it's an internal function that's only called in about five or six places in this file, I decided to leave it alone and just fix up the res path after the fact.  But maybe it really _would_ make sense to have `init_resb_result()` just take two parameters for these values; I can take a shot at that if you want.

One interesting this is that the two spots where we want these values to be different were doing two things: Here, we're setting both to the "key" value and throwing away the additional path elements.  In the similar spot in ICU-20400, we were setting both to the "path" value.  That didn't seem to be hurting anything, since nobody in that code path was really looking at the "key" value-- it mainly made debugging more difficult.  But it was also scary because the resource bundle doesn't own the storage for its `fKey` member-- sticking a path in there only worked because we happened to be setting it to point to another resource's `fResPath` member (which _does_ own its storage).  Of course, if the other resource happened to go away before this one did, we'd be in trouble, but that wasn't happening in practice (Thank God).

Let me know if you want to take a whack at just having `init_resb_result()` take separate parameters for the key and path; that seems doable.  If you don't think it's worth the trouble, I'll squash.

Reply:
Not worth now I think, but sounds fishy. Please submit another ticket with something like what you wrote here. We should really take some time (pair programming) looking through the code and the API semantics and thinking about what makes sense.

As I said elsewhere, sometimes we get into trouble because we have "who am I" APIs and the answer sometimes depends on what you did or what you want to use the answer for. (Mark likes to add "who am I" APIs, and I have been resisting for some time now.)

Reply:
Okay, I've filed https://unicode-org.atlassian.net/browse/ICU-21715 for the additional work, and I'll go ahead and squash this.  Thank you!